Add overview of custom NGF policies (#2088)
nginxinc/nginx-gateway-fabric: site/content/overview/custom-policies.md

| ClientSettingsPolicy                                                 | Configure connection behavior between client and NGINX  | Inherited       | Gateway, HTTPRoute, GRPCRoute | No                            | Yes       | v1alpha1    |
| [ObservabilityPolicy]({{<relref "/how-to/monitoring/tracing.md" >}}) | Define settings related to tracing, metrics, or logging | Direct          | HTTPRoute, GRPCRoute          | Yes                           | No        | v1alpha1    |

{{< important >}}
NGINX Gateway Fabric policies do not work with [HTTPRoute matches](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1.HTTPRouteMatch) with `headers`, `params`, or `method` matchers defined. This will be addressed in a future release.
{{< /important >}}

## Terminology

- _Attachment Type_. How the policy attaches to an object. Attachment type can be "direct" or "inherited".
- _Supported Target Object(s)_. API objects the policy can be applied to.
- _Supports Multiple Target Refs_. Whether a single policy can target multiple objects.
- _Mergeable_. Whether policies that target the same object can be merged.

## Direct Policy Attachment

> A Direct Policy Attachment is a policy that references a single object -- such as a Gateway or HTTPRoute. It is tightly bound to one instance of a particular Kind within a single Namespace or an instance of a single Kind at the cluster-scope. It affects _only_ the object specified in its TargetRef.

Reviewer: ```suggestion
A Direct Policy Attachment is a policy that references a single object, such as a Gateway or HTTPRoute. It is tightly bound to one instance of a particular Kind within a single Namespace or an instance of a single Kind at the cluster-scope. It affects _only_ the object specified in its TargetRef.
```

This diagram uses a fictional retry policy to show how Direct Policy Attachment works:

> {{<img src="img/direct-policy-attachment.png" alt="Direct Policy Attachment">}}

Reviewer: ```suggestion
{{<img src="img/direct-policy-attachment.png" alt="">}}
```
Alternative text is meant to be a replacement for an image, not a title stating what it is.

An empty alt tag has specific meaning to a browser (Distinctly different to a "missing" one): the alt text needs to be a complete description of the image or the paragraph underneath must include all of the relevant information.

The policy targets the HTTPRoute `baz` and sets `retries` to `3` and `timeout` to `60s`. Since this policy is a Direct Policy Attachment, its settings are only applied to the `baz` HTTPRoute.

## Inherited Policy Attachment

> Inherited Policy Attachment is designed to allow settings to flow down a hierarchy. The hierarchy for Gateway API resources looks like this:

Reviewer: ```suggestion
Inherited Policy Attachment allows settings to cascade down a hierarchy. The hierarchy for Gateway API resources looks like this:
```

> {{<img src="img/hierarchy.png" alt="Hierarchy">}}

Reviewer: ```suggestion
{{<img src="img/hierarchy.png" alt="">}}
```

See previous note about image alt text.

Settings defined in a policy attached to an object in this hierarchy may be inherited by the resources below it. For example, the settings defined in a policy attached to a Gateway may be inherited by all the HTTPRoutes attached to that Gateway.

Settings in an Inherited Policy can be Defaults or Overrides. Defaults set the default value for something and can be overridden by policies on a lower object. Overrides cannot be overridden by lower objects.
All settings in NGINX Gateway Fabric Inherited Policies are Defaults.

Default values are given precedence from the bottom up. Therefore, a policy setting attached to a Backend will have the highest precedence over the one attached to higher objects.

The following diagram shows how Inherited Policies work in NGINX Gateway Fabric using a fictional retry policy:

> {{<img src="img/inherited-policy-attachment.png" alt="Inherited Policy Attachment">}}

Reviewer: ```suggestion
{{<img src="img/inherited-policy-attachment.png" alt="">}}
```

See previous note about image alt text.

There are three policies defined:

- Blue policy that targets the `dev` Gateway
- Orange policy that targets the `baz` HTTPRoute

> - Pink policy that targets the `foo` HTTPRoute

Reviewer 2: Not sure our policy on including colors in the docs (thinking about color-blindness).

Author: Good call. I'll name the policies instead

With some NGINX Gateway Fabric Policies, it is possible to create multiple policies that target the same resource as long as the fields in those policies do not conflict.

For example, consider the following fictional policies:

```yaml
apiVersion: gateway.nginx.org/v1alpha1
kind: ExamplePolicy
metadata:
  name: retries
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: HTTPRoute
    name: foo
  retries: 10
```

```yaml
apiVersion: gateway.nginx.org/v1alpha1
kind: ExamplePolicy
metadata:
  name: timeout
spec:
  targetRef:
    kind: HTTPRoute
    name: foo
  timeout: 60s
```

The `retries` ExamplePolicy defines the number of retries for the `foo` HTTPRoute, and the `timeout` ExamplePolicy defines the timeout for the `foo` HTTPRoute.
NGINX Gateway Fabric will merge the fields defined in the policies and apply the following settings to the `foo` HTTPRoute:

```yaml
retries: 10
timeout: 60s
```

However, if both policies had the `retries` field set, then the policies cannot be merged. In this case, NGINX Gateway Fabric will choose which policy to configure based on the following criteria (continuing on ties):

1. The oldest policy by creation timestamp

> 2. The policy appearing first in alphabetical order by "{namespace}/{name}"

Reviewer: ```suggestion
1. The policy appearing first in alphabetical order by "{namespace}/{name}"
```

Hugo will automatically enumerate list items starting with 1. This feels unnecessary for 2 items, but means ordered lists of arbitrary size can be changed without requiring each item to be manually numbered.

## Policy Status

NGINX Gateway Fabric sets the [PolicyStatus](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1alpha2.PolicyStatus) on all policies.

`PolicyStatus` fields:

- `ancestors`: describes the status of a route with respect to the ancestor.
  - `ancestorRef`: the object that the policy targets in `spec.targetRef`
  - `controllerName`: the controller name of NGINX Gateway Fabric.
  - `conditions`: (Condition/Status/Reason):
    - `Accepted/True/Accepted`: the policy is accepted by the ancestor.
    - `Accepted/False/Invalid`: the policy is not accepted because it is semantically or syntactically invalid.
    - `Accepted/False/Conflicted`: the policy is not accepted because it conflicts with another policy.
    - `Accepted/False/TargetNotFound`: the policy is not accepted because it targets a resource that is invalid or does not exist.
    - `Accepted/False/NginxProxyNotSet`: the policy is not accepted because it relies on the NginxProxy configuration which is missing or invalid.

> 

Reviewer: ```suggestion
```

    kind: HTTPRoute
    name: foo
  timeout: 60s
```

The `retries` ExamplePolicy defines the number of retries for the `foo` HTTPRoute, and the `timeout` ExamplePolicy defines the timeout for the `foo` HTTPRoute.
NGINX Gateway Fabric will merge the fields defined in the policies and apply the following settings to the `foo` HTTPRoute:

```yaml
retries: 10
timeout: 60s
```

However, if both policies had the `retries` field set, then the policies cannot be merged. In this case, NGINX Gateway Fabric will choose which policy to configure based on the following criteria (continuing on ties):

1. The oldest policy by creation timestamp
1. The policy appearing first in alphabetical order by "{namespace}/{name}"

If a policy conflicts with a configured policy, NGINX Gateway Fabric will set the policy `Accepted` status to false with a reason of `Conflicted`. See [Policy Status](#policy-status) for more details.

## Policy Status

NGINX Gateway Fabric sets the [PolicyStatus](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1alpha2.PolicyStatus) on all policies.

`PolicyStatus` fields:

- `ancestors`: describes the status of a route with respect to the ancestor.

>   - `ancestorRef`: the object that the policy targets in `spec.targetRef`

Reviewer 3: ```suggestion
  - `ancestorRef`: the object that the policy targets in `spec.targetRef`.
```
